rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,227 @@
+software versions and protocol versions that correlate to that release. This
+will reside in `spec/version.md`. A clear guide of the rules behind breaking
+changes should also be there for easy reference.
+
+### Implementation Approach
+
+If this proposal is accepted, the implementation will need to be further fleshed
+out in an ADR, however, it is helpful to illustrate here the approach that
+Tendermint would most likely take to support soft upgrades.
+
+This would mostly likely mimic a similar design to what has been done in the
+Cosmos SDK with keeping legacy code and using conditionals to handle transitions
+between the versions. One would expect a directory structure for each module as
+such:
+
+```go
+consensus // module
+|---- state.go
+|---- state_test.go
+|       ...
+|---- votes.go // current version (v10)
+|       ...
+|---- legacy
+   |---- v8 // prior supported versions
+      |---- votes.go
+      |      ...    // other data structures ( and pure functions )
+   |---- v9
+      |---- votes.go
+      |      ...
+```
+
+The module, depending on the types, may statically cast each of the different
+versions or define an interface that it expects all versions to follow. For
+example:
+
+```go
+func (blockExec *BlockExecutor) CreateProposalBlock(state State) (BlockI, error)
+{
+ switch state.Version.Block {
+ case 9:
+  return v9.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ case 10:
+  return v10.CreateProposalBlock(state, blockExec.Mempool, blockExec.Evpool)
+ default:
+  return fmt.Errorf("block version %d not recognised", state.Version.Block)
+ }
+}
+```
+
+In this case, Block which falls under a data structure of the block protocol
+could have many versions and is represented here as an interface. `State` and
+`BlockExecutor` are structs outside of the block protocol. Breaking changes to
+them could only occur in major releases.
+
+### Alternative Approach: Migration Scripts
+
+Although using migration scripts would have been most likely a simpler solution,
+unfortunately in most cases this would result in breaking the hashes and is
+therefore not a viable approach.
+
+### Concluding Remarks
+
+This option provides greater flexibility for the developers and potentially
+reduces the amount of coordination and work required for validators and the
+frequency with which hard forks need to occur. However, it is nevertheless
+recommended that block protocol breaking changes still be grouped together and
+remain as seldom as possible.
+
+The implementation to support this would ideally need to be done prior to the
+1.0 release.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Chains get the benefits of upgrades without having to sacrifice their block
+history
+
+### Negative
+
+- There is a large degree of complexity required in ensuring the flexibility of
+data structures. This could lead to the bloating of the code-base and/or
```

Comment:
Putting different version of Tendermint into a tm-version repo could mitigate this. This way tendermint only holds the latest versions and avoids bloating. 

There is some added complexity to this but keeps the repo clean and minimal.

Reply:
and just import it as a library? Yeah that could be an alternative implementation

Reply:
> and just import it as a library? Yeah that could be an alternative implementation

Yea!

Reply:
could you import from the same repo but from different releases?

Reply:
Potentially. Once we make the jump to proper semvar then we will want to look at how go modules recommends versioning: https://blog.golang.org/v2-go-modules. 

```md
Major versions and module paths

Modules formalized an important principle in Go, the import compatibility rule:

If an old package and a new package have the same import path,
the new package must be backwards compatible with the old package.

By definition, a new major version of a package is not backwards compatible with the previous version. This means a new major version of a module must have a different module path than the previous version.
```

Although this section conflicts with allowing block protocol breaking in an X.X release.